lr_finder.py

```diff
@@ -282,3 +275,40 @@ def get_lr(self):
         curr_iter = self.last_epoch + 1
         r = curr_iter / self.num_iter
         return [base_lr * (self.end_lr / base_lr) ** r for base_lr in self.base_lrs]
+
+
+class StateCacher(object):
+    def __init__(self, in_memory, cache_dir=None):
+        self.in_memory = in_memory
+        self.cache_dir = cache_dir
+
+        if self.cache_dir is None:
+            import tempfile
+            self.cache_dir = tempfile.gettempdir()
+        else:
+            if not os.path.isdir(self.cache_dir):
+                raise ValueError('Given `cache_dir` is not a valid directory.')
+
+        self.cached = {}
+
+    def store(self, key, state_dict):
+        if self.in_memory:
+            self.cached.update({key: copy.deepcopy(state_dict)})
+        else:
+            fn = os.path.join(self.cache_dir, 'state_{}.pt'.format(key))
+            self.cached.update({key: fn})
+            torch.save(state_dict, fn)
+
+    def retrieve(self, key):
+        if key not in self.cached:
+            raise KeyError('Target {} was not cached.'.format(key))
+
+        if self.in_memory:
+            return self.cached.pop(key)
```

Comment:
If the cache is popped, `reset` can only be called one time; that doesn't allow the user to run multiple range tests with different settings and compare results because after the first call the model and optimizer cannot be reset to the initial state again.

Reply:
You are right, I missed that point.
I'll change `self.cached.pop(...)` to `self.cached.get(...)`.

Besides, I found that we should let `StateCacher` to check whether there still are files not being used by `StateCacher.retrieve()` before a `StateCacher` instance is going to be destroyed.
I think that we can do this check in `StateCacher.__del__()`.